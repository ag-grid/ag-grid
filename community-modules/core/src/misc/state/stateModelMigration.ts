import type { GridState } from '../../interfaces/gridState';

export function migrateGridStateModel(state: GridState): GridState {
    // The `version` field was introduced in v32.2.0, so anything without that
    // field can be assumed to be compatible with v32.1.0
    if (!state.version) {
        state.version = '32.1.0';
    }

    switch (state.version) {
        case '32.1.0':
            state = migrateV32_1(state);
    }

    return state;
}

function migrateV32_1(state: GridState): GridState {
    state.cellSelection = jsonGet(state, 'rangeSelection');
    jsonDelete(state, 'rangeSelection');
    return state;
}

function jsonGet(json: unknown, key: string): any {
    if (json && typeof json === 'object') {
        return (json as any)[key];
    }
}

function jsonDelete(json: unknown, key: string): void {
    if (json && typeof json === 'object') {
        delete (json as any)[key];
    }
}
