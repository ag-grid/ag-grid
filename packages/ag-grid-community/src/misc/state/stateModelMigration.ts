import type { GridState } from '../../interfaces/gridState';

export function migrateGridStateModel(state: GridState): GridState {
    state = { ...state };
    // The `version` field was introduced in v32.2.0, so anything without that
    // field can be assumed to be compatible with v32.1.0
    if (!state.version) {
        state.version = '32.1.0';
    }

    switch (state.version) {
        case '32.1.0':
            state = migrateV32_1(state);
    }

    return state;
}

/**
 * Copies `rangeSelection` (deprecated) to `cellSelection`, but doesn't remove the former
 * for backwards compatibility
 */
function migrateV32_1(state: GridState): GridState {
    state.cellSelection = jsonGet(state, 'rangeSelection');
    return state;
}

function jsonGet(json: unknown, key: string): any {
    if (json && typeof json === 'object') {
        return (json as any)[key];
    }
}
