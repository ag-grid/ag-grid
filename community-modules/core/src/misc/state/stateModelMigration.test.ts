import fs from 'fs';
import path from 'path';

import type { GridState } from '../../interfaces/gridState';
import { migrateGridStateModel } from './stateModelMigration';

const FIXTURES_DIR = path.join(__dirname, 'fixtures');

function loadFixtures(): [name: string, version: string | undefined, state: GridState][] {
    return fs
        .readdirSync(FIXTURES_DIR)
        .filter((f) => f.endsWith('.json'))
        .map((file) => {
            const contents = JSON.parse(fs.readFileSync(path.join(FIXTURES_DIR, file), 'utf-8')) as GridState;
            return [file, contents.version, contents];
        });
}

describe('Grid State Migration', () => {
    test.each(loadFixtures())('%s: should upgrade version %s', (name, version, state) => {
        const migrated = migrateGridStateModel(state);
        expect(migrated).toMatchSnapshot();
    });
});
