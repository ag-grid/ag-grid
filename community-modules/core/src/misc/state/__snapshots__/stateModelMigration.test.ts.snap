// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid State Migration range-selection-state.json: should upgrade version undefined 1`] = `
{
  "cellSelection": {
    "cellRanges": [],
  },
  "sort": {
    "sortModel": [],
  },
  "version": "32.1.0",
}
`;
